src/main/java/de/tum/in/www1/artemis/service/TimeBasedAchievementService.java

```diff
@@ -29,6 +32,38 @@ public TimeBasedAchievementService(AchievementRepository achievementRepository)
         this.achievementRepository = achievementRepository;
     }
 
+    /**
+     * Checks whether a user earned time based achievements
+     * @param result the result which is checked if it earned any achievements
+     * @param achievements all time based achievements within the given course
+     * @return the highest rank reached by the result, returns null if no rank was reached
+     */
+    public AchievementRank checkForAchievement(Result result, Set<Achievement> achievements) {
+        if (result.getScore() == null || !achievements.iterator().hasNext() || result.getScore() < achievements.iterator().next().getMinScoreToQualify()) {
+            return null;
+        }
+        var submission = result.getSubmission();
+        if (submission == null) {
+            return null;
+        }
+
+        var submissionDay = submission.getSubmissionDate().truncatedTo(DAYS);
+        var exerciseReleaseDay = result.getParticipation().getExercise().getReleaseDate().truncatedTo(DAYS);
+
+        Set<AchievementRank> ranks = new HashSet<>();
+
+        for (Achievement achievement : achievements) {
+            if (submissionDay.minusDays(achievement.getSuccessCriteria()).isEqual(exerciseReleaseDay)
+                    || submissionDay.minusDays(achievement.getSuccessCriteria()).isBefore(exerciseReleaseDay)) {
```

Comment:
You can shorten this if you like:
```suggestion
            if (submissionDay.minusDays(achievement.getSuccessCriteria()+1).isBefore(exerciseReleaseDay)) {
```